### DISCUSSION TITLE
Month Validation with Arrow?

### DISCUSSION THREAD

**Discussion by Swizzler121** [crsmithdev/arrow#1071] [Q&A]
I'm trying to replace datetime in my python, everything so far has worked beautifully, but I have an argument parser that I have a line that checks an input month to see if it's valid. the arrow version works fine if a user uses a 2 digit month, but if you use a single digit month, it says it's not valid. the datetime version converts single digit months to 2 digit before validation, so it works fine if a user gives a single or 2 digit month. I was hoping arrow could go even further and convert a written month to a 2 digit month, but I can't figure out how to do this.

the datetime variant:

`parser.add_argument("-m","--month", type=lambda d: datetime.strptime(d, '%m'), help="specify a month (integer)")`

**Comment by systemcatch** [comment]
Hi @Swizzler121, if I'm understanding your question correctly then here's how to do what you want.

Use the `M` token to parse single digit months.

```shell
>>> s="2020-9-17"
>>> arrow.get(s, "YYYY-M-DD")
<Arrow [2020-09-17T00:00:00+00:00]>
```

https://arrow.readthedocs.io/en/latest/#supported-tokens

For the written month you could convert like this.

```shell
>>> dt=arrow.get("March", "MMMM")
>>> dt
<Arrow [0001-03-01T00:00:00+00:00]>
>>> dt.month
3
>>> dt.format("MM")
'03'
```

arrow also wraps `strparse` if that helps.

```shell
>>> arrow.Arrow.strptime("2", "%m")
<Arrow [1900-02-01T00:00:00+00:00]>
```

**Reply by Swizzler121** [reply]
When I try to drop in arrow where datetime was with .strptime it throws an error:

`int() argument must be a string, a bytes-like object or a number, not 'Arrow'`

```py
parser.add_argument(
	"-m","--month", 
	nargs='+', 
	type=lambda d: arrow.Arrow.strptime(d, '%m'),
	help="specify a month (integer)"
)
```

is the line in question. the lambda is there to check if a month is valid. works with datetime, but fails with arrow.

This works, but the lambda month-checking doesn't work, if you don't feed it a 2-digit month it errors out:
```py
parser.add_argument(
	"-m","--month", 
	nargs='+', 
	type=lambda d: arrow.get(d, 'MM').format('MM'), 
	help="specify a month (integer)"
)
```

**Comment by systemcatch** [comment]
Can you provide the full error message output?

Try doing this.
```python
parser.add_argument(
	"-m","--month", 
	nargs='+', 
	type=lambda d: arrow.get(d, 'MM').format('MM').datetime, 
	help="specify a month (integer)"
)
```

**Comment by Swizzler121** [comment]
The full error before:
```
~$python script.py -m 8

usage: ezp-stats.py [-h] [-y YEAR [YEAR ...]] [-m MONTH [MONTH ...]]
ezp-stats.py: error: argument -m/--month: invalid <lambda> value: '8'
```

The error now:
```
~$python script.py -m 8

  File "/home/git/EZproxy_LogFileAnalysis/ezp-stats.py", line 122, in <lambda>
    type=lambda d: arrow.get(d, 'MM').format('MM').datetime, 
AttributeError: 'str' object has no attribute 'datetime'
```

Just to reiterate, the reason I'm feeding it 1 digit dates instead of two is I want the system to be able to interpret both, and maybe even written months if possible, The datetime version I posted in my first post takes a 1 or 2 digit number, checks if its between 1-12, then if it is passes it to the program, if not, it errors out. The arrow version works if the month values are 2 digits.

**Comment by systemcatch** [comment]
Ahh sorry I gave you the wrong code, try this.

```python
parser.add_argument(
	"-m","--month", 
	nargs='+', 
	type=lambda d: arrow.Arrow.strptime(d, '%m').datetime, 
	help="specify a month (integer)"
)
```

This should give you exactly the same return value for `type` as your original code.

**Reply by Swizzler121** [reply]
That works, but still doesn't support single-digit month dates. I'm not sure what datetime is doing different to handle conversion from single digit to 2 digits. 

I might just relent and break it into it's own function to handle all possible cases. I wish I could use match but I'm trying to keep it compatible with 3.7+ for now.

The good news is that showed me how to fix another issue where I had to use the datetime import to get the week# starting from sunday since your week# formatting has a bunch of extra info, and also counts weeks starting on mondays. So while that didn't get my code working like it did with datetime in one place, it did in another.

It would be a cool bit of functionality to be able to use %m to check all permutations of month arrow supports in the future though. I appreciate what your team is trying to do with the library, it helped slim down my project a ton where I didn't have to sit there and fight limitations with timedelta so much.

**Reply by Swizzler121** [reply]
actually, scratch that, the code doesn't work. It works for a single argument, but when giving multiple arguments it fails with error:
```
  File "/home/chrg/git/EZproxy_LogFileAnalysis/ezp-stats.py", line 214, in set_stat_range
    s = arrow.get().replace(year=int(y[0]),month=int(m[0]))
TypeError: int() argument must be a string, a bytes-like object or a number, not 'datetime.datetime'
```

replacing `type=lambda d: arrow.Arrow.strptime(d, '%m').datetime, ` with `type=lambda d: arrow.get(d, 'MM').format('MM'), ` resolves the issue.
I'm not sure if this issue would have occurred with the datetime library as well, as when I was using datetime I wasn't dealing with ranges as large as I am now

The range I used was: `python ezp-stats.py -y 2019 2020 -m 10 10`
which pulls a stat-file range from october 2019 to october 2020.

**Reply by systemcatch** [reply]
It would almost certainly occur with `datetime` as this is a `TypeError` due to the wrong type of object being passed to the `int` function.

Without seeing your full code it is hard to understand what is actually going on.

The code given will parse single and two digit months fine.

```shell
>>> arrow.Arrow.strptime("2", '%m').datetime
datetime.datetime(1900, 2, 1, 0, 0, tzinfo=tzutc())
>>> arrow.Arrow.strptime("02", '%m').datetime
datetime.datetime(1900, 2, 1, 0, 0, tzinfo=tzutc())
>>> arrow.Arrow.strptime("11", '%m').datetime
datetime.datetime(1900, 11, 1, 0, 0, tzinfo=tzutc())
```
